src/Opaleye/TH.hs

```diff
@@ -0,0 +1,652 @@
+        let deleteFunc = FunD (mkName "deleteModel") [Clause pat (NormalB deleteExp) []]
+        return [InstanceD Nothing [] (AppT (ConT $ mkName "DbModel") (ConT $ mkName r)) [insertFunc, updateFunc, deleteFunc]]
+      where
+        makeConversionFunction :: Q [Dec]
+        makeConversionFunction = do
+          let
+            pgReadType = ConT $ (mkName $ makePGReadTypeName r)
+            pgWriteType = ConT $ (mkName $ makePGWriteTypeName r)
+            conversionFunctionSig = SigD (mkName "toWrite") $ AppT (AppT ArrowT pgReadType) pgWriteType
+            conversionFunction = FunD (mkName "toWrite") [Clause [makePattern] (NormalB conExp) []]
+          return $ [conversionFunctionSig, conversionFunction]
+          where
+            conExp :: Exp
+            conExp = foldl AppE (ConE $ mkName $ r) $ getFieldExps
+            getFieldExps :: [Exp]
+            getFieldExps = zipWith getFieldExp fieldInfos [1..]
+              where
+              getFieldExp :: ColumnInfo -> Int -> Exp
+              getFieldExp ci ix = if (columnDefault ci) then (AppE (ConE $ mkName "Just") mkVarExp) else mkVarExp
+                where
+                  mkVarExp :: Exp
+                  mkVarExp = VarE $ mkName $ "a" ++ (show ix)
+            makePattern :: Pat
+            makePattern = ConP (mkName $ r) $ VarP <$> (mkName.(\x -> ('a':x)).show) <$>  take (Prelude.length fieldInfos) [1..]
+        getPrimaryKeyField :: String -> String -> (Maybe String)
+        getPrimaryKeyField tname modelName = case (filter (\ci -> columnPrimary ci)) fieldInfos of
+            [primaryField] -> Just $ makeFieldName modelName (columnName primaryField)
+            _ -> Nothing
+  
+makeTablename :: String -> String
+makeTablename t = t ++ "Table"
+
+makePGReadTypeName :: String -> String
+makePGReadTypeName tn = tn ++ "PGRead"
+
+makePGWriteTypeName :: String -> String
+makePGWriteTypeName tn = tn ++ "PGWrite"
+
+makeAdapterName :: String -> String
+makeAdapterName tn = 'p':tn
+
+makeOpaleyeModels :: (ConnectInfo, Options) -> Q [Dec]
+makeOpaleyeModels (ci, op) = makeOpaleyeModels' (ci, op, [])
+
+makeOpaleyeModels' :: Env -> Q [Dec]
+makeOpaleyeModels' env = fst <$> runStateT (do
+  newTypeDecs <- makeNewTypes
+  (_, options, _) <- get
+  let names = fst <$> tableOptions options
+  let models = (modelName.snd) <$> tableOptions options
+  instances <- makeNewtypeInstances
+  decs <- concat <$> zipWithM makeOpaleyeModel names models
+  return $ newTypeDecs ++ decs ++ instances) env
+
+collectNewTypes :: EnvM [(ColumnInfo, String)]
+collectNewTypes = do
+  (connInfo, options, _) <- get
+  concat <$> mapM getNewTypes (tableOptions options)
+  where
+    getNewTypes :: (String, TableOptions) -> EnvM [(ColumnInfo, String)]
+    getNewTypes (tbName, tbOptions) = do
+      (connectInfo, _, _) <- get
+      fieldInfos <- (lift.runIO) $ do
+        conn <- connect connectInfo
```

Comment:
is this going to re-connect for every table? Pass an open connection in the `Env` along with the connect info, perhaps?